pkg/metricshandler/metrics_handler.go

```diff
@@ -0,0 +1,257 @@
+			if err != nil {
+				klog.Errorf("detect sharding settings from StatefulSet: %v", err)
+				return
+			}
+
+			m.mtx.RLock()
+			shardingUnchanged := m.curShard == shard && m.curTotalShards == totalShards
+			m.mtx.RUnlock()
+
+			if shardingUnchanged {
+				return
+			}
+
+			m.ConfigureSharding(ctx, shard, totalShards)
+		},
+		UpdateFunc: func(oldo, curo interface{}) {
+			old := oldo.(*appsv1.StatefulSet)
+			cur := curo.(*appsv1.StatefulSet)
+			if cur.Name != statefulSetName {
+				return
+			}
+
+			if old.ResourceVersion == cur.ResourceVersion {
+				return
+			}
+
+			shard, totalShards, err := shardingSettingsFromStatefulSet(cur, m.opts.Pod)
+			if err != nil {
+				klog.Errorf("detect sharding settings from StatefulSet: %v", err)
+				return
+			}
+
+			m.mtx.RLock()
+			shardingUnchanged := m.curShard == shard && m.curTotalShards == totalShards
+			m.mtx.RUnlock()
+
+			if shardingUnchanged {
+				return
+			}
+
+			m.ConfigureSharding(ctx, shard, totalShards)
+		},
+	})
+	go i.Run(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), i.HasSynced) {
+		return errors.New("waiting for informer cache to sync failed")
+	}
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+// ServeHTTP implements the http.Handler interface. It writes the metrics in
+// its stores to the response body.
+func (m *MetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+	resHeader := w.Header()
+	var writer io.Writer = w
+
+	resHeader.Set("Content-Type", `text/plain; version=`+"0.0.4")
+
+	if m.enableGZIPEncoding {
+		// Gzip response if requested. Taken from
+		// github.com/prometheus/client_golang/prometheus/promhttp.decorateWriter.
+		reqHeader := r.Header.Get("Accept-Encoding")
+		parts := strings.Split(reqHeader, ",")
+		for _, part := range parts {
+			part = strings.TrimSpace(part)
+			if part == "gzip" || strings.HasPrefix(part, "gzip;") {
+				writer = gzip.NewWriter(writer)
+				resHeader.Set("Content-Encoding", "gzip")
+			}
+		}
+	}
+
+	for _, s := range m.stores {
+		s.WriteAll(w)
+	}
+
+	// In case we gzipped the response, we have to close the writer.
+	if closer, ok := writer.(io.Closer); ok {
+		closer.Close()
+	}
+}
+
+func shardingSettingsFromStatefulSet(ss *appsv1.StatefulSet, podName string) (nominal int32, totalReplicas int, err error) {
+	nominal, err = detectNominalFromPod(ss.Name, podName)
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "detecting Pod nominal")
+	}
+
+	totalReplicas = 1
+	replicas := ss.Spec.Replicas
+	if replicas != nil {
+		totalReplicas = int(*replicas)
+	}
+
+	return nominal, totalReplicas, nil
+}
+
+func detectNominalFromPod(statefulSetName, podName string) (int32, error) {
+	nominalString := strings.TrimPrefix(podName, statefulSetName+"-")
+	nominal, err := strconv.Atoi(nominalString)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to detect shard index for Pod %s of StatefulSet %s, parsed %v", podName, statefulSetName, nominalString)
```

Comment:
Let's use the string directive here.
```suggestion
		return 0, errors.Wrapf(err, "failed to detect shard index for Pod %s of StatefulSet %s, parsed %s", podName, statefulSetName, nominalString)
```